store/zookeeper/zookeeper.go

```diff
@@ -66,19 +68,29 @@ func (s *Zookeeper) setTimeout(time time.Duration) {
 // Get the value at "key", returns the last modified index
 // to use in conjunction to Atomic calls
 func (s *Zookeeper) Get(key string) (pair *store.KVPair, err error) {
-	resp, meta, err := s.client.Get(s.normalize(key))
 
-	if err != nil {
-		if err == zk.ErrNoNode {
-			return nil, store.ErrKeyNotFound
+	var resp []byte
+	var meta *zk.Stat
+
+	// To guard against older versions of libkv
+	// creating and writing to znodes non-atomically, 
+	// We try to resync few times if we read SOH or 
+	// an empty string
+	for i := 0; i <= syncRetryLimit; i++ {
+		resp, meta, err = s.client.Get(s.normalize(key))
+
+		if err != nil {
+			if err == zk.ErrNoNode {
+				return nil, store.ErrKeyNotFound
+			}
+			return nil, err
 		}
-		return nil, err
-	}
 
-	// FIXME handle very rare cases where Get returns the
-	// SOH control character instead of the actual value
-	if string(resp) == SOH {
-		return s.Get(store.Normalize(key))
+		if (string(resp) == SOH || string(resp) == "") && i < syncRetryLimit {
```

Comment:
The i < syncRetryLimit is to prevent a Sync in the very last iteration of the for loop. Without this, we would unnecessarily call sync and then exit the loop. (So effectively we call sync not more than syncRetryLimit number of times).